### DISCUSSION TITLE
"sh: node-waf: command not found" after upgrade to npm@7

### DISCUSSION THREAD

**Discussion by Richie765** [verdaccio/verdaccio#2316] [Q&A]
I've recently upgraded npm from 6 to 7. Now I get the following `npm install ` error:

```
npm ERR! code 127
npm ERR! path /Users/richard/tmp/quasar-app/node_modules/zlib
npm ERR! command failed
npm ERR! command sh -c node-waf clean || true; node-waf configure build
npm ERR! sh: node-waf: command not found
npm ERR! sh: node-waf: command not found
```

I'm doing `quasar create quasar-app` which has a dependency for `zlib`.

When I rename my `~/.npmrc` to something else, it works fine (so it goes directly to npm without verdaccio).
Also when I go back to npm@6 it works fine.

I've tried cleaning npm cache. I've also tried on other computers (Mac and Linux). I also upgraded to the latest verdaccio. The problem stays the same.

It seems upgrading to npm@7 introduces this issue, but without Verdaccio it also works fine. I'm now back to npm@6 but can't stay there forever and I have custom packages in Verdaccio. Any idea what could solve it?

**Comment by Richie765** [comment]
v14.17.1
I know but it works when I bypass verdaccio.

**Reply by Richie765** [reply]
I tried with:
node v16.4.1
npm 7.19.1

Same error. When I move .npmrc out of the way it works.

**Reply by DanielRuf** [reply]
That would be weird. Because according to package.json zlib is always loaded from npmjs, which is not correct.

You probably have a prebuilt version of zlib on your local machine. Also verdaccio has its own cache folder. So `npm i --force` would probably refetch and rebuild all packages and their dependencies, including zlib.

Besides this, the zlib reference in package.json is not a good idea. This should be fixed by the quasar developers.

**Reply by Richie765** [reply]
I was thinking about a prebuild package as well. That could explain why it works in certain cases. But why can't it find it then? 
I am using my own ca-cert (cafile=xxx in .npmrc). Could this be a reason?

**Reply by DanielRuf** [reply]
You can try `npm i --loglevel verbose` or `npm i --loglevel silly` to see what happens in the background.

**Reply by Richie765** [reply]
I don't see any other error. With .npmrc out of the way, I don't see that it is compiling anything. I also don't see that it downloads a precompiled binary (at least I cannot find it inside node_modules). Anyways it's a bit a mystery to me why it works in certain cases and not in the other. For now I'll go back to npm@6 and ask quasar later.

**Reply by DanielRuf** [reply]
Probably the other projects use a different version of quasar. The change was introduced last year in https://github.com/quasarframework/quasar/commit/3a41bba0eb1445996c823b1f42ddc0eefa95168b

With `npm ls` you can see the dependencies of a project, for example `npm ls zlib` will filter by this dependency.

It seems there is now a PR which should resolve this in future releases: https://github.com/quasarframework/quasar/pull/9951

**Reply by Richie765** [reply]
Yes, the PR was created in response of my issue, so great to see this is picked up so quickly! Hope it will pass through soon.

However it's not due to a different Quasar version as it is the exact same project. The only differences is in npm version and using/not-using verdaccio:

With npm@7 + verdaccio it doesnt work.
With npm@6 + verdaccio it works
With npm@7 without verdaccio it works.

So this remains a strange to me.

**Reply by DanielRuf** [reply]
The loglevel will show you, if a package is fetched and compiled or if a precompiled version / binding is found in the cache path and used.

But so far without any additional logs, we can only assume and guess, what the exact difference is.

@juanpicado are you aware of any difference introduced by verdaccio? Because so far this makes not much sense to me. Nothing runs `configure` in `node_modules/zlib`.

**Reply by juanpicado** [reply]
No, also does not make much sense to me. The main problem I see here is try to debug in such big scope (quasar framework), to prove the point I'd reduce to a simple `package.json` project to prove the issue is on verdaccio. The 95% of the time the issue is external.

Anyway, verdaccio does not unpack anything or modify files in anyway, just stream zip content from the origin to the client. 

One hint issue might be external is works with other package managers,  npm6 is not that different to npm7 in terms of communication with verdaccio, @Richie765 you can try with yarn and pnpm as well.

@Richie765 please try to provide a small reducible step.

Remember, even switching `nvm` node versions to use different `npm` the content is being chached in your home directory anyway, so I'd also recommend reproduce in a Dockerfile where also ensure a clean state.

**Reply by Richie765** [reply]
I appreciate you guys looking into it. I've investigated all I could including the detailed logs, but I couldn't find anything that caused this. I'm not blaming Verdaccio per-se just that it works fine when I bypass it. I also discovered another minor potential issue with Verdaccio but I'll leave that for another time. 

I think it's best if you try to reproduce the issue yourself, since you know better what to look at. So here are the exact steps.

* I'm using a fairly clean and up-to date Ubuntu 20.10, node, etc are all clean installed.
* Using 'n' to install node
* The following does 4x "quasar create". The 2 without verdaccio, the last 2 with verdaccio. Once with npm@6 and once with npm@7.
* In the end I'll demonstrante clearly that it works without verdaccio but fails with
* I have marked where the interesting part starts, you can skip to there if you like
* I tried many things (different versions, etc) but the results are consistent

```bash
node --version
# v14.17.2
npm --version
6.14.13

# install quasar
npm install --global @quasar/cli

# try with npm 6 - no verdaccio
cat ~/.npmrc
# file not found

mkdir ~/tmp
cd ~/tmp
quasar create my-app
# hit enter's and choose "npm"
# Installation succeeds
# cleanup
rm -rf my-app
rm -rf ~/.npm

# try with npm 7 - no verdaccio
cat ~/.npmrc
# file not found
npm install --global npm@7
cd ~/tmp
quasar create my-app
# hit enter's and choose "npm"
# Installation succeeds
# cleanup
rm -rf my-app
rm -rf ~/.npm

## VERDACCIO STARTS HERE

npm set registry https://verdaccio.mydomain/
npm set cafile ~/local-ca.crt
npm login

# try with npm 6 - WITH verdaccio
cat ~/.npmrc
# file exists, has: registry, cafile and correct authToken
npm install --global npm@6
cd ~/tmp
quasar create my-app
# hit enter's and choose "npm"
# Installation succeeds
# cleanup
rm -rf my-app
rm -rf ~/.npm

##
## HERE STARTS THE INTERESTING PART
##

# try with npm 7 - WITH verdaccio
cat ~/.npmrc
# file exists, has: registry, cafile and correct authToken
npm install --global npm@7
cd ~/tmp
quasar create my-app
# hit enter's and choose "npm"
# Installation FAILS
# cleanup
rm -rf my-app
rm -rf ~/.npm

# bonus - remove ~/.npmrc (bypass verdaccio) and try once again
mv ~/.npmrc ~/.npmrc-off
quasar create my-app
# SUCESS
# put ~/.npmrc back again
mv ~/.npmrc-off ~/.npmrc
rm -rf my-app
quasar create my-app
# FAILS AGAIN
```

Error
```
npm ERR! code 127
npm ERR! path /home/richard/tmp/my-app/node_modules/zlib
npm ERR! command failed
npm ERR! command sh -c node-waf clean || true; node-waf configure build
npm ERR! sh: 1: node-waf: not found
npm ERR! sh: 1: node-waf: not found
```

**Comment by DanielRuf** [comment]
For me it seems the zlib package (on npmjs.com) uses a very old command. node-waf was used before node-gyp, which is the current solution to build native modules.

zlib is already an internal package of nodejs, see https://nodejs.org/api/zlib.html

In my opinion the quasar framework should not try to load zlib from npmjs, see https://github.com/quasarframework/quasar/blob/4daa1975cc74d49424289bc5eb389a3318addedf/app/package.json#L116

So I would suggest to report this issue to the quasar developers, because zlib is in node since the beginning (see https://nodejs.org/docs/latest-v0.10.x/api/zlib.html).

**Reply by Richie765** [reply]
Good idea, I'll file an issue there.

**Reply by DanielRuf** [reply]
Also see https://github.com/kkaefer/DEPRECATED-node-zlib/blob/master/Makefile and https://github.com/kkaefer/DEPRECATED-node-zlib/blob/master/configure, so the `node-waf` command is always coming from the zlib package, except you already have a locally built version of it somewhere in the PATH variable.
